samza-core/src/main/java/org/apache/samza/table/caching/CachingTable.java

```diff
@@ -105,96 +102,222 @@ public void init(SamzaContainerContext containerContext, TaskContext taskContext
     tableMetricsUtil.newGauge("req-count", () -> requestCount());
   }
 
+  /**
+   * Lookup the cache and return the keys that are missed in cache
+   * @param keys keys to be looked up
+   * @param records result map
+   * @return list of keys missed in the cache
+   */
+  private List<K> lookupCache(List<K> keys, Map<K, V> records) {
+    List<K> missKeys = new ArrayList<>();
+    records.putAll(cache.getAll(keys));
+    keys.forEach(k -> {
+        if (!records.containsKey(k)) {
+          missKeys.add(k);
+        }
+      });
+    return missKeys;
+  }
+
   @Override
   public V get(K key) {
+    try {
+      return getAsync(key).get();
+    } catch (InterruptedException e) {
+      throw new SamzaException(e);
+    } catch (Exception e) {
+      throw (SamzaException) e.getCause();
+    }
+  }
+
+  @Override
+  public CompletableFuture<V> getAsync(K key) {
     readMetrics.numGets.inc();
-    long startNs = System.nanoTime();
     V value = cache.get(key);
-    if (value == null) {
-      missCount.incrementAndGet();
-      Lock lock = stripedLocks.get(key);
-      try {
-        lock.lock();
-        if (cache.get(key) == null) {
-          // Due to the lack of contains() API in ReadableTable, there is
-          // no way to tell whether a null return by cache.get(key) means
-          // cache miss or the value is actually null. As such, we cannot
-          // support negative cache semantics.
-          value = rdTable.get(key);
-          if (value != null) {
-            cache.put(key, value);
-          }
-        }
-      } finally {
-        lock.unlock();
-      }
-    } else {
+    if (value != null) {
       hitCount.incrementAndGet();
+      return CompletableFuture.completedFuture(value);
     }
-    readMetrics.getNs.update(System.nanoTime() - startNs);
-    return value;
+
+    long startNs = System.nanoTime();
+    missCount.incrementAndGet();
+
+    return rdTable.getAsync(key).handle((result, e) -> {
+        if (e != null) {
+          throw new SamzaException("Failed to get the record for " + key, e);
+        } else {
+          if (result != null) {
+            cache.put(key, result);
+          }
+          readMetrics.getNs.update(System.nanoTime() - startNs);
+          return result;
+        }
+      });
   }
 
   @Override
   public Map<K, V> getAll(List<K> keys) {
+    try {
+      return getAllAsync(keys).get();
+    } catch (InterruptedException e) {
+      throw new SamzaException(e);
+    } catch (Exception e) {
+      throw (SamzaException) e.getCause();
```

Comment:
Problem with we always get ExecutionException with CompletableFuture.get() and the execution fails. In our handle method for the async future, we already wrap any IO exception with a SamzaException, which becomes the cause for the ExecutionException. So instead of having layers of nested exceptions, we simply propagate the cause.  

I think this is cleaner. What do you think?

Reply:
getAsync and get are two methods neck-to-neck in the same class soI don't fully agree with the coupling argument. But I don't have a strong opinion of nested exceptions, I will wrap it with a new SamzaException then.